File: base/src/main/java/me/shetj/base/base/BaseFragment.java
package me.shetj.base.base;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Keep;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.trello.rxlifecycle2.components.support.RxAppCompatActivity;
import com.trello.rxlifecycle2.components.support.RxFragment;

import org.simple.eventbus.EventBus;

import java.lang.reflect.Field;

import me.shetj.base.tools.app.ArmsUtils;
import me.shetj.base.view.LoadingDialog;

/**
 * fragment基类
 * @author shetj
 */
@Keep
public abstract class BaseFragment<T extends BasePresenter> extends RxFragment implements IView {
    private static final String STATE_SAVE_IS_HIDDEN = "STATE_SAVE_IS_HIDDEN";

    /**
     * The M activity.
     */
    protected Context mActivity;
    /**是否可见状态*/
    protected boolean isVisible;
    /**View已经初始化完成*/
    private boolean isPrepared;
    /**是否第一次加载完*/
    private boolean isFirstLoad = true;

    protected T mPresenter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        isFirstLoad = true;
        //绑定View
        isPrepared = true;
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initEventAndData();
        lazyLoad();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mActivity = getActivity();
        //如果要使用eventbus请将此方法返回true
        if (useEventBus())
        {
	        EventBus.getDefault().register(this);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (useEventBus())
        {
	        EventBus.getDefault().unregister(this);
        }
        this.mActivity = null;
    }

    /**
     * 是否使用eventBus,默认为使用(true)，
     *
     * @return boolean
     */
    protected boolean useEventBus() {
        return true;
    }


    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putBoolean(STATE_SAVE_IS_HIDDEN, isHidden());
    }

    @Override
    public void onAttach(Context context) {
        this.mActivity = context;
        super.onAttach(context);
    }
    @Override
    public void onDetach() {
        super.onDetach();
        try {
            Field childFragmentManager = Fragment.class.getDeclaredField("mChildFragmentManager");
            childFragmentManager.setAccessible(true);
            childFragmentManager.set(this, null);
        } catch (NoSuchFieldException e) {
            throw new RuntimeException(e);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if(getUserVisibleHint()){
            isVisible = true;
            onVisible();
        }else {
            isVisible = false;
            onInvisible();
        }
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if(!hidden){
            isVisible = true;
            onVisible();
        }else {
            isVisible = false;
            onInvisible();
        }
    }
    /**
     * On visible.
     */
    protected void onVisible(){
        lazyLoad();
    }

    /**
     * On invisible.
     */
    protected void onInvisible(){}

    /**
     * Lazy load.
     */
    protected void lazyLoad(){
        if(!isPrepared || !isVisible || !isFirstLoad) {
	        return;
        }
        isFirstLoad = false;
        lazyLoadData();
    }

    /**
     * Init event and data.
     */
    protected abstract void initEventAndData();

    /**
     * Lazy load data.
     */
    public abstract void lazyLoadData();


    @Override
    public void showLoading(String msg) {

        LoadingDialog.showLoading(getRxContext(), msg, true);
    }
    @Override
    public void hideLoading() {
       LoadingDialog.hideLoading();
    }

    @Override
    public void showMessage(@NonNull String message) {
        ArmsUtils.makeText(message);
    }

    /**
     * 返回当前的activity
     * @return
     */
    @Override
    public RxAppCompatActivity getRxContext(){
        return (RxAppCompatActivity) mActivity;
    }

    @Override
    public void onDestroyView() {
        if (mPresenter != null) {
            mPresenter.onDestroy();
        }
        super.onDestroyView();
    }
    @SuppressLint("unchecked")
    @Override
    public void updateView(BaseMessage message) {

    }
}

File: base/src/main/java/me/shetj/base/base/BaseFragment.kt
package me.shetj.base.base

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.os.Message
import androidx.annotation.Keep
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.trello.rxlifecycle3.components.support.RxAppCompatActivity
import com.trello.rxlifecycle3.components.support.RxFragment

import org.simple.eventbus.EventBus

import me.shetj.base.tools.json.EmptyUtils
import me.shetj.base.tools.json.GsonKit
import timber.log.Timber

/**
 * fragment基类
 * @author shetj
 */
@Keep
abstract class BaseFragment<T : BasePresenter<*>> : RxFragment(), IView {

    /**
     * The M activity.
     */
    protected var mActivity: Context? = null
    /**是否可见状态 */
    protected var isShow: Boolean = false
    /**View已经初始化完成 */
    private var isPrepared: Boolean = false
    /**是否第一次加载完 */
    private var isFirstLoad = true

    protected var mPresenter: T? = null

    /**
     * 返回当前的activity
     * @return RxAppCompatActivity
     */
    override val rxContext: RxAppCompatActivity
        get() = (mActivity as RxAppCompatActivity?)!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isFirstLoad = true
        //绑定View
        isPrepared = true
        initEventAndData()
        lazyLoad()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mActivity = activity
        //如果要使用eventbus请将此方法返回true
        if (useEventBus()) {
            EventBus.getDefault().register(this)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (useEventBus()) {
            EventBus.getDefault().unregister(this)
        }
        this.mActivity = null
    }

    /**
     * 是否使用eventBus,默认为使用(true)，
     *
     * @return boolean
     */
    protected fun useEventBus(): Boolean {
        return true
    }


    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(STATE_SAVE_IS_HIDDEN, isHidden)
    }

    override fun onAttach(context: Context) {
        this.mActivity = context
        super.onAttach(context)
    }

    override fun onDetach() {
        super.onDetach()
        try {
            val childFragmentManager = Fragment::class.java.getDeclaredField("mChildFragmentManager")
            childFragmentManager.isAccessible = true
            childFragmentManager.set(this, null)
        } catch (e: NoSuchFieldException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }

    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (userVisibleHint) {
            isShow = true
            onVisible()
        } else {
            isShow = false
            onInvisible()
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            isShow = true
            onVisible()
        } else {
            isShow = false
            onInvisible()
        }
    }

    /**
     * On visible.
     */
    protected fun onVisible() {
        lazyLoad()
    }

    /**
     * On invisible.
     */
    protected fun onInvisible() {}

    /**
     * Lazy load.
     */
    protected fun lazyLoad() {
        if (!isPrepared || !isShow || !isFirstLoad) {
            return
        }
        isFirstLoad = false
        lazyLoadData()
    }

    /**
     * Init event and data.
     */
    protected abstract fun initEventAndData()

    /**
     * Lazy load data.
     */
    fun lazyLoadData() {

    }


    override fun onDestroyView() {
        if (mPresenter != null) {
            mPresenter!!.onDestroy()
        }
        super.onDestroyView()
    }

    @SuppressLint("unchecked")
    override fun updateView(message: Message) {
        if (EmptyUtils.isNotEmpty(message)) {
            Timber.i(GsonKit.objectToJson(message))
        }
    }

    companion object {
        private const val STATE_SAVE_IS_HIDDEN = "STATE_SAVE_IS_HIDDEN"
    }
}
